pydantic_ai_slim/pydantic_ai/prompt_config.py

```diff
@@ -0,0 +1,414 @@
+            tool return confirmations, and validation error messages.
+        tool_config: Configuration for customizing tool descriptions and metadata.
+    """
+
+    templates: PromptTemplates | None = None
+    """Templates for customizing system-generated messages sent to the model.
+
+    See [`PromptTemplates`][pydantic_ai.PromptTemplates] for available template options.
+    """
+
+    tool_config: dict[str, ToolConfig] | None = None
+    """Configuration for customizing tool descriptions and metadata, keyed by tool name.
+    See [`ToolConfig`][pydantic_ai.ToolConfig] for available configuration options.
+    """
+
+    @staticmethod
+    async def generate_prompt_config_from_agent(agent: Agent[Any, Any], model: Model) -> PromptConfig:
+        """Generate a PromptConfig instance based on an Agent instance.
+
+        This fills in per-tool metadata and default templates, producing a `PromptConfig` that can be
+        used as a starting point for prompt optimizers.
+
+        """
+        tool_config: dict[str, ToolConfig] = {}
+
+        prompt_templates: PromptTemplates = PromptTemplates(
+            final_result_processed=DEFAULT_FINAL_RESULT_PROCESSED,
+            output_tool_not_executed=DEFAULT_OUTPUT_TOOL_NOT_EXECUTED,
+            output_validation_failed=DEFAULT_OUTPUT_VALIDATION_FAILED,
+            function_tool_not_executed=DEFAULT_FUNCTION_TOOL_NOT_EXECUTED,
+            tool_denied=DEFAULT_TOOL_CALL_DENIED,
+            validation_errors_retry_prompt=DEFAULT_MODEL_RETRY,
+            tool_retry_prompt=DEFAULT_MODEL_RETRY,
+            no_tool_retry_prompt=DEFAULT_MODEL_RETRY,
+            prompted_output_template=DEFAULT_PROMPTED_OUTPUT_TEMPLATE,
+            description_template=None,
+        )
+
+        run_ctx = RunContext(deps=None, model=model, usage=RunUsage())
+
+        # Include both regular and output tools
+        from .toolsets import CombinedToolset
+
+        all_toolsets = [*agent.toolsets]
+        if output_toolset := getattr(agent, '_output_toolset', None):
+            all_toolsets.append(output_toolset)
+
+        toolset = CombinedToolset(all_toolsets)
+        tools = await toolset.get_tools(run_ctx)
+
+        for tool_name, toolset_tool in tools.items():
+            tool_def = toolset_tool.tool_def
+            tool_config[tool_name] = ToolConfig(
+                name=tool_name,
+                tool_description=tool_def.description,
+                strict=tool_def.strict,
+                tool_args_descriptions=_extract_descriptions_from_json_schema(tool_def.parameters_json_schema),
+            )
+
+        return PromptConfig(
+            tool_config=tool_config,
+            templates=prompt_templates,
+        )
+
+
+# JSON Schema keys
+_PROPERTIES = 'properties'
+_DEFS = '$defs'
+_REF = '$ref'
+_REF_PREFIX = '#/$defs/'
+_DESCRIPTION = 'description'
+
+
+def _extract_descriptions_from_json_schema(parameters_json_schema: dict[str, Any]) -> dict[str, str]:
```

Comment:
I'd like this to live closer to the code for injecting descriptions into the schema. Maybe a new class with 2 methods? Or 2 utils? Either way, side by side.